agent/connect/ca/aws_pca.go

```diff
@@ -0,0 +1,438 @@
+			return NewAmazonPCA(client, config, newARN, pcaType, keyAlgorithm, signingAlgorithm, logger), nil
+		}
+
+		logger.Print("[DEBUG] sleeping until certificate is ready")
+		time.Sleep(5 * time.Second) // TODO: get from provider config
+	}
+}
+
+func (pca *AmazonPCA) describe() (*acmpca.DescribeCertificateAuthorityOutput, error) {
+	input := &acmpca.DescribeCertificateAuthorityInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+	}
+
+	return pca.client.DescribeCertificateAuthority(input)
+}
+
+func (pca *AmazonPCA) GetCSR() (string, error) {
+	input := &acmpca.GetCertificateAuthorityCsrInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+	}
+	pca.logger.Printf("[DEBUG] retrieving CSR for %s", pca.arn)
+	output, err := pca.client.GetCertificateAuthorityCsr(input)
+	if err != nil {
+		pca.logger.Printf("[ERR] error retrieving CSR: %s", err.Error())
+		return "", err
+	}
+
+	return *output.Csr, nil
+}
+
+func (pca *AmazonPCA) SetCert(certPEM string, chainPEM string) error {
+	chainBytes := []byte(chainPEM)
+	if chainPEM == "" {
+		chainBytes = nil
+	}
+	input := acmpca.ImportCertificateAuthorityCertificateInput{
+		CertificateAuthorityArn: aws.String(pca.arn),
+		Certificate:             []byte(certPEM),
+		CertificateChain:        chainBytes,
+	}
+
+	pca.logger.Printf("[DEBUG] uploading certificate for %s", pca.arn)
+	_, err := pca.client.ImportCertificateAuthorityCertificate(&input)
+	if err != nil {
+		pca.logger.Printf("[ERR] error importing certificates: %s", err.Error())
+		return err
+	}
+
+	pca.certPEM = certPEM
+	pca.chainPEM = chainPEM
+
+	return nil
+}
+
+func (pca *AmazonPCA) getCerts() error {
+	if pca.certPEM == "" || pca.chainPEM == "" {
+		input := &acmpca.GetCertificateAuthorityCertificateInput{
+			CertificateAuthorityArn: aws.String(pca.arn),
+		}
+
+		output, err := pca.client.GetCertificateAuthorityCertificate(input)
+		if err != nil {
+			return err
+		}
+
+		pca.certPEM = *output.Certificate
+		pca.chainPEM = *output.CertificateChain
+	}
+
+	return nil
+}
+
+func (pca *AmazonPCA) Certificate() string {
+	if pca.certPEM == "" {
+		_ = pca.getCerts()
+	}
+	return pca.certPEM
+}
+
+func (pca *AmazonPCA) CertificateChain() string {
+	if pca.chainPEM == "" {
+		_ = pca.getCerts()
+	}
+	return pca.certPEM
+}
+
+func (pca *AmazonPCA) Generate(signingPCA *AmazonPCA) error {
+	csrPEM, err := pca.GetCSR()
+	if err != nil {
+		return err
+	}
+
+	templateARN := GetTemplateARN(pca.pcaType)
+	chainPEM := ""
+	validity := RootValidity
+	if pca.pcaType == acmpca.CertificateAuthorityTypeSubordinate {
+		chainPEM = signingPCA.Certificate()
+		validity = IntermediateValidity
+	}
+
+	newCertPEM, err := signingPCA.Sign(csrPEM, templateARN, validity)
+	if err != nil {
+		return err
+	}
+
+	err = pca.SetCert(newCertPEM, chainPEM)
+	return err
+}
+
+func (pca *AmazonPCA) Sign(csrPEM string, templateARN string, validity time.Duration) (string, error) {
+	issueInput := acmpca.IssueCertificateInput{
```

Comment:
In original TLS RFC  (CSL-029 I think) we specified a bunch of validations to perform on the CSR. The one I can recall is that we need to assert that the SpiffeID being claimed by the CSR is a valid Consul SPIFFE ID and that it has the same trust domain as the current cluster ID (so we can't be tricked into signing arbitrary certs for a different trust domain that might share a CA root).

Can you check the other providers and end ensure we have equivalent validation of the CSR? It _might_ be that the CA code already does some/all of that at a higher level but we should be absolutely sure!

Reply:
The other providers don't do much validation either. The Vault provider just PEM encodes the CSR and immediately sends it to Vault. The Consul provider checks the SPIFFE ID in the CSR to be sure it's a valid SPIFFE ID.